ggml backends interface, ggml-cuda refactor

slaren opened the pull request:
Previous discussion: https://github.com/ggerganov/llama.cpp/pull/2230

This PR adds a common interface to the compute backends.

### Breaking changes
- `ggml_context` allocates memory from a `ggml_buffer` that contains a buffer in device memory for the tensor data, and a buffer in system memory for the tensor structs (and in the future also other data such as the graphs)
- The `data` member of `ggml_tensor` is a backend-specific pointer that should not be accessed directly. To access the data, `ggml_backend_set_tensor` and `ggml_backend_get_tensor` must be used instead. Functions such as `ggml_new_f32` and `ggml_set_f32` can also be used as before.
  - Technically, if you are using the CPU backend, you can still access the `data` member directly, but you shouldn't do that if you want to support other backends
  - I will probably change the name to something else to prevent current code from compiling without changes
- Added a small `params` buffer to `ggml_tensor` for the op parameters that currently are stored in a tensor. For example, for `ggml_rope` this buffer is used to store the values `n_past, n_dims, mode, n_ctx`. The goal is to make these parameters easily accessible from the CPU, and reduce the overhead of creating a new tensor for them.

Brief example:
```C
// initialize a backend
struct ggml_backend backend = ggml_backend_cpu_init();
// for CUDA:
// struct ggml_backend backend = ggml_backend_cuda_init();

// create a buffer
size_t ctx_size = 4096; // buffer size for the tensor data
size_t num_tensors = 10; // maximum number of tensors that can be allocated
struct ggml_buffer buf = ggml_backend_alloc_buffer(&backend, ctx_size, num_tensors);

// create a context using the buffer
struct ggml_init_params ggml_params = ggml_init_params_default();
ggml_params.buffer = &buf;
ggml_context * ctx = ggml_init(ggml_params);

// use the context to create a computation graph
struct ggml_tensor * x = ggml_new_f32(ctx, 2.0f);
struct ggml_tensor * a  = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, 1);
struct ggml_tensor * b  = ggml_new_tensor_1d(ctx, GGML_TYPE_F32, 1);
struct ggml_tensor * x2 = ggml_mul(ctx, x, x);
struct ggml_tensor * f  = ggml_add(ctx, ggml_mul(ctx, a, x2), b);
struct ggml_cgraph gf = ggml_build_forward(f);

// set the value of the input tensors
float a_val = 3.0f;
ggml_backend_set_tensor(a, &a_val, 0, sizeof(float));
// alternatively:
float b_val = 5.0f;
ggml_set_f32(b, b_val);

// run the computation
ggml_backend_graph_compute(&backend, &gf);

// get the result
float result;
ggml_backend_get_tensor(f, &result, 0, sizeof(float));
// alternatively:
// result = ggml_get_f32_1d(f, 0);
```

### Backend implementation
Backends should implement the functions defined in the `ggml_backend_interface` struct. Currently there are implementations for the CPU and CUDA backends.

### Computation using multiple backends
It is still possible to offload some parts of the graph to the GPU while keeping others on the CPU. This is done using `ggml_graph_splits`. See the llama.cpp code for an example, will update this later with more details.

### Notes/limitations
- Only the CPU and CUDA backends are currently supported.
- Only the bare minimum necessary to run llama is currently implemented, don't look too much into the details of the code for now
- When partially offloading a model to the GPU, internally this is handled by splitting the graphs into multiple parts, and running each of them in sequence in a different backend. Because the KV memory is either on the CPU or on the GPU, this means that there are at least as many graph executions as there are layers. The CPU backend creates and destroys the threads with every graph launch, so this can have a non-negligible overhead. Eventually this will be improved by using a thread pool.
- On the plus side, the CPU threads are no longer spinning while the GPU backend is running
- In the long term, the goal is to support automatic fallback to the CPU when an operation is not implemented in a backend, but that's going to take a while. This means that backends like OpenCL that mostly only implement matrix multiplication cannot be used for now. I am expecting that this backend will be replaced by a Vulkan backend that implements full GPU offloading, and won't be necessary anymore.
- For prompt processing, it is usually worth to upload the weights every time even when using partial offloading, but that's currently not supported. So prompt processing will use the same CPU/GPU layer split as generation. Eventually I would like to support this again.

slaren committed:
ggml backend interface wip

refactor ggml-cuda

slaren committed:
restore simple.cpp for now

slaren committed:
minor fixes

slaren committed:
fix NVCC version on Makefile, __halves2half2 -> make_half2

slaren committed:
rebase to master (except ggml-cuda)

Dampfinchen commented:
This sentence concerns me a bit: "For prompt processing, it is usually worth to upload the weights every time even when using partial offloading, but that's currently not supported. So prompt processing will use the same CPU/GPU layer split as generation. Eventually I would like to support this again."

Does that mean when the PR is merged, people using partial offloading (like me running only 10 layers on the GPU with a 13B model) will be experiencing significant longer prompt evaluation times?

If that's the case, perhaps it would be better to wait until the performance is atleast on the same level before merging. IMO LLama.cpp's most important feature that sets it apart from GPTQ and BNB is the possibility to run models too large for the hardware at great performance, it would be sad to lose that.

Then of course, there is the concern from AMD and Intel GPU owners. OpenCl has been an important backend for them to get decent speed on their hardware. So they would not benefit anymore from CPU performance improvements and newer quantization format in newer releases if this PR gets merged. Which would be a shame. Please keep in mind the Vulkan backend might take a long time to develop as it is a very complex task. If you are eager to supersede OpenCL in such a short time, perhaps you could lend a hand!

Aside from these concerns, I am very impressed by the awesome work you and others are doing and I fully support this PR! I'm just pointing out the effects it may have on people enjoying the software.

slaren commented:
I understand your concerns. I will add OpenCL support again in the same way that it works currently. Once we get closer to merging this, I will run some tests to see exactly what is the performance impact on the prompt processing speed and re-evaluate from there.

Don't get me wrong, I agree that these features are important, and I want to continue to support them. The issue is that PRs like this that affect a lot of code can have a significant maintenance overhead, because other changes merged in the meanwhile will often cause conflicts that must be resolved manually. Ultimately, it is a matter of whether we want to prioritize development speed or maintaining a stable master branch.

casperbh96 commented:
Weighing in on the issue of breaking things versus shipping this feature faster. 

I weigh heavily towards shipping faster and fixing things after. Developers should in general lock in on a commit (GPT4All does this) before this PR is merged so that they have a stable version until performance is stabilized afterwards and bugs are squashed. 

The problem with trying to cater to solving these problems like slower prompt processing or CPU offloading is that it hinders long-term progress for a short-term benefit.

On the plus side, this PR will enable a future where more models can more easily be supported due to the backend setup. The faster it’s shipped, the faster we get performance optimization and generally better tech to support our LLMs.

slaren committed:
only allocate as much memory as is required in each backend for the model

slaren commented on ggml-backend.h:
```diff
@@ -0,0 +1,129 @@
+    static inline void          ggml_backend_alloc_tensor(struct ggml_buffer * buffer, struct ggml_tensor * tensor) { buffer->backend->interface->alloc_tensor(buffer->backend->context, buffer->backend_buffer, tensor); }
+
+    // tensor copy between different backends
+    GGML_API void ggml_backend_cpy_tensor(struct ggml_tensor * dst, struct ggml_tensor * src);
```
The order of the arguments of `ggml_backend_cpy_tensor` is the opposite of `ggml_cpy`, which can be confusing. However, this is the same order as functions such as `memcpy`, so I am not sure what is better.

ggerganov replied:
Let's keep the order as `ggml_cpy()` - first `src` then `dst`

We also have these:

https://github.com/ggerganov/llama.cpp/blob/d01bccde9f759b24449fdaa16306b406a50eb367/ggml.h#L1540-L1546

slaren committed:
improve layer backend printing with ranges

slaren committed:
fix double-free with --no-mmap

slaren committed:
allocators wip
renamed ggml_backend functions
changed ggml_buffer and ggml_backend to always be used as pointers
rename ggml_tensor::params -> op_params

slaren committed:
llama.cpp: remove backend-specific code where possible

slaren committed:
automatically calculate compute buffer sizes (without graph allocator)

slaren committed:
initial implementation of delayed graph allocation

slaren committed:
improved memory management

slaren committed:
remove unused code

slaren committed:
llama.cpp: partially restore state support, graph export

slaren committed:
improved memory management fixes

slaren commented:
@JohannesGaessler how confident are you in the VRAM scratch buffer size estimation for context sizes > 2048? I have been working on https://github.com/ggerganov/ggml/issues/288 and I get these sizes for the compute buffer (n_batch = 512):

| n_ctx | compute buffer (calculated) | CUDA VRAM scratch buffer (master) |
--- | --- | --- |
| 512 | 88.50 MB | 288 MB|
| 1024 | 152.00 MB | 320 MB|
| 2048 | 280.00 MB | 384 MB|
| 4096 | 536.00 MB | 512 MB|
| 8192 | 1048.00 MB | 768 MB|
| 16384 | 2072.00 MB | 1280 MB|

For smaller context sizes, the calculated memory is smaller (which is expected as this should be more efficient than scratch buffers), but for larger contexts it is bigger. It is likely that there are bugs in my implementation, though, but perplexity seems fine.

ggerganov commented:
@slaren 

I believe that @ikawrakow has also observed an issue with the currently estimated VRAM sizes for larger contexts (https://github.com/ggerganov/llama.cpp/pull/2295 I haven't looked at details yet, so I could be wrong). But I guess your sizes might be actually the correct ones.

JohannesGaessler commented:
See https://github.com/ggerganov/llama.cpp/pull/2056 . I tested the VRAM scratch sizes with a precision of 1 MB, meaning I determined the lowest scratch size that still produces correct results. I tested up to 8192 context if possible. I observed that the amount of VRAM required increases linearly with context. I chose the VRAM scratch sizes on master to be the minimum required scratch size +25% at 2048 context. It's possible that the implementation on master is suboptimal and thus needs more VRAM for that reason.

slaren commented:
Thank you. It is not clear how you estimated the memory, it may be possible that we are using different parameters. What I do is construct a graph with these parameters and calculate the size of the compute buffer required to execute it:
```c++
int n_tokens = std::min((int)hparams.n_ctx, params.n_batch);
int n_past = hparams.n_ctx - n_tokens;
```
This should be equivalent to computing a prompt of `n_batch` tokens at the latest possible position in the context window, which I think is the worst case.

slaren commented:
I understand that, what I am wondering about is how you ran the tests exactly. If what you did is different than this, then different results are expected:
>This should be equivalent to computing a prompt of n_batch tokens at the latest possible position in the context window, which I think is the worst case.

JohannesGaessler commented:
If I remember correctly I was calculating perplexity on the first 100 lines of Wikitext as a test.

slaren commented:
Ok, that should be equivalent when `n_ctx` is a multiple of `n_batch`, so I looked what tensors were allocated at the peak memory usage and I found that it was mostly `KQ` and `KQ_scaled`. Previously `KQ_scaled` was inplace, but I had removed inplace operations. So I added automatic inplace operations to the allocator, and I get these results now:
| n_ctx | compute buffer (calculated) | VRAM scratch buffer|
--- | --- | --- |
| 512 | 70.50 MB | 288 MB|
| 1024 | 88.00 MB | 320 MB|
| 2048 | 152.00 MB | 384 MB|
| 4096 | 280.00 MB | 512 MB|
| 8192 | 536.00 MB | 768 MB|
| 16384 | 1048.00 MB | 1280 MB|

So it should be a reduction in memory usage over master in all cases.

slaren committed:
allocator: automatic inplace operations

JohannesGaessler commented on ggml-backend.c:
```diff
@@ -705,32 +708,83 @@ void allocate_graph(struct ggml_cgraph * gf, struct ggml_buffer * buffer) {
 
-void ggml_graph_allocate_tensors_n(struct ggml_cgraph ** graphs, int n_graphs) {
-}
+static bool ggml_is_view(struct ggml_tensor * t) {
```
Is this name intentional? It seems confusing considering the collision with `GGML_OP_VIEW`.

slaren replied:
I use view generally to mean any operation that shares the memory of its parent tensor. In think the name makes sense, but I am open to suggestions.

JohannesGaessler commented on ggml-backend.c:
```diff
@@ -815,6 +866,32 @@ void allocate_node(struct ggml_buffer * buffer, struct ggml_tensor * node) {
+                    break;
+                }
+                // TODO: make a list of operations that can be safely made inplace
+                if (parent->data != NULL && parent->n_children == 1 && parent->n_views == 0 && ggml_are_same_layout(node, parent) && node->op != GGML_OP_MUL_MAT) {
```
It's not used in LLaMA but `GGML_CONV_2D` should also not be done inplace.

slaren committed:
allocator: fix partial offloading

slaren committed:
allocator cleanup

slaren committed:
llama.cpp: print input/output buffers size

slaren commented:
@ggerganov or anyone familiar with the training code, is there any way to tell what tensors must be kept for the backwards pass? maybe checking if `ggml_tensor::grad` is not NULL? I am not sure how to rewrite this after removing the scratch buffers:

https://github.com/ggerganov/llama.cpp/blob/24baa54ac1ff3d4156a2360deb1473af04a9b1a2/examples/train-text-from-scratch/train-text-from-scratch.cpp#L1604-L1634

slaren committed:
allocator: cleanup, more comments

ggerganov commented:
If you keep all tensors with `grad != NULL` you will be able to compute the backwards pass.
However, this is not optimal as it will keep pretty much everything from the forward pass, while depending on the operations, the values of some tensors are not needed.

For example, let's take a look at the backward pass for `GGML_OP_ADD`:

https://github.com/ggerganov/llama.cpp/blob/24baa54ac1ff3d4156a2360deb1473af04a9b1a2/ggml.c#L15187-L15195

```bash
# forward
tensor = z

slaren commented:
If it is determined manually, I could add a parameter to the allocator to specify the list of tensors that must not be freed. That should achieve the same goal of minimizing memory usage, while still keeping the data of the important tensors. It could possibly be a list in `ggml_cgraph`, let's call it `outputs` to represent the list of outputs of a graph.

ggerganov commented:
Ok, we can do that for now - it should be better than the current approach, thought it might be possible to improve it even further in the future to do everything automatically. Remind me, at what point does the allocator run - during `ggml_build_forward()`?

slaren commented:
Currently, it is called manually, but the idea is it should be done automatically after building the graph. Essentially, when you create a context, you can specify the tensor allocation mode. For the graph allocator (`GGML_ALLOC_COMPUTE_SEQ`), this should happen after building the graph in `ggml_build_forward`.
https://github.com/ggerganov/llama.cpp/blob/d273bfd2c98148fab309f978c44ec2e8c24d1c4d/llama.cpp#L1193-L1197

slaren commented:
Closing this as it is too outdated to continue from this point.
